### DISCUSSION TITLE
Attempting to get snapshot by any means only yields defaultValues of atoms

### DISCUSSION THREAD

**Discussion by nobleach** [facebookexperimental/Recoil#1527] [Q&A]
I'm looking for a way to dump the state of the application. The docs specify that a Snapshot is the correct way to do this. I've tried all methods: `useRecoilSnapshot`, `useRecoilCallback` and `snapshot_UNSTABLE()`. None of them seem to be able to dump my updated value. 

I've created a codesandbox to demonstrate: https://codesandbox.io/s/determined-wing-vtddn?file=/src/App.js:777-796

Overall, I'm working toward dumping the state so I can rehydrate a `<RecoilRoot />`.  I'd at least like to prove that I can dump the most up to date version of the state.

**Comment by drarmstr** [comment]
The `snapshot_UNSTABLE()` call creates a fresh snapshot that is not associated with the `<RecoilRoot>` state, so the call to `setProductId()` will never effect it.

You shouldn't call functions like `setProductId()` or `logState()` with side-effects from a React render function directly.  Something like a `useEffect()` would be more appropriate.  Note that `setProductId()` enqueues the state change, but it may not occur synchronously with the call.  (Though, it is observable immediately from subsequent updaters.)  That's why the initial snapshots still have the default value of `9`.  Though, note that the subsequent callback snapshot from `logState()` does reflect the updated value of `5`.

**Comment by nobleach** [comment]
>The snapshot_UNSTABLE() call creates a fresh snapshot that is not associated with the `<RecoilRoot>`

Ok, this makes sense. I appreciate the clarification. So If I'm understanding you correctly, Recoil has no way of dumping the current state of the application with values? 

> You shouldn't call functions like `setProductId()` or `logState()`

Absolutely agree. This was a quick codesandbox to learn how this all worked.

**Comment by drarmstr** [comment]
Yes, you should be able to get a snapshot of the current state with `useRecoilCallback()` or `useRecoilSnapshot()`.  The snapshot you got from the callback was updated, try updating the state from an effect to see if that resolves your issue with having the snapshot from `useRecoilSnapshot()` reflect the updated state as well.

**Comment by nobleach** [comment]
Yes, using `useEffect` does indeed yield the expected result. Since this code will be run during a server render, `useEffect` will not run. Since `snapshot_UNSTABLE()` runs outside of the context of the `<RecoilRoot />` I'm guessing it'll never be able to acquire a snapshot of the updated store?

Basically, I'm working to find if there's any way I can get SSR working. I realize that's out of scope of Recoil presently... I've just seen musings here and there that it may be possible with snapshots. So far I've been able to dump snapshots with default values... but there's really no path for rehydrating without taking all the data and resetting all atoms on the client side.

**Comment by drarmstr** [comment]
> Yes, using `useEffect` does indeed yield the expected result. Since this code will be run during a server render, `useEffect` will not run. Since `snapshot_UNSTABLE()` runs outside of the context of the `<RecoilRoot />` I'm guessing it'll never be able to acquire a snapshot of the updated store?
> 
> 
> 
> Basically, I'm working to find if there's any way I can get SSR working. I realize that's out of scope of Recoil presently... I've just seen musings here and there that it may be possible with snapshots. So far I've been able to dump snapshots with default values... but there's really no path for rehydrating without taking all the data and resetting all atoms on the client side.

useRecoilSnapshot() will return a Snapshot for the state of the current rendering. Recoil does support initializing state to a non-default state for the initial rendering for SSR.  You can initialize atoms using Atom Effects or the initializeState prop of <RecoilRoot>.

**Comment by nobleach** [comment]
Yeah, that `initializeState` prop takes a function that is passed a MutableSnapshot and returns void, right? I was looking at how that might work.

The docs specify that that prop takes a function.... but it's a function that doesn't return anything. So, if I have an instance of a Snapshot, I can't imagine that I just pass that.... would that work?

**Comment by drarmstr** [comment]
So you already have a `Snapshot` instance when you are trying to render?  I guess `useGotoRecoilSnapshot()` wouldn't work if you need this for an initial server rendering.  The `initializeState` prop can't just take the Snapshot as-is, since it's a mutable snapshot callback, as you point out.  Though, I suppose the interface could be updated to also accept a `Snapshot`.  Otherwise, you should be able to just manually iterate though the atoms in your snapshot and set them in the `initializeState` callback.

**Reply by nobleach** [reply]
Yes. My thought is, if I could serialize the Snapshot to JSON, attach it to the `window` object, and when the browser boots the app, deserialize it and feed it to the `<RecoilRoot />` (the tried and true method we've been doing with Redux for years).

The issue with manually iterating through the snapshot and setting all those values is, that work has already been done during that initial server render. Manually iterating means I have to dig around for all those values (that are coming from several different sources... API calls, etc). As you can imagine, this could get very messy. BUT, if that's the current solution, that's what I'll do.

Regarding the snapshot itself. It doesn't appear serializable (at least to JSON). It contains functions, etc. 

Feeding an already initialized (and full of set values) Snapshot could be a killer feature, BUT, if that Snapshot can't be serialized, it makes things a bit more tricky. (Not that one couldn't implement `toJSON()` on the Snapshot class, and use _that_ as a potential value for the `initializeState` prop.

**Reply by drarmstr** [reply]
The `Snapshot` object itself isn't JSON serializable.  You could iterate the nodes and serialize the values, as long as the values you use are serializable.  For manually iterating you don't have to dig around, you can just iterate using the `getNodes_UNSTABLE()` method of the `Snapshot`.  My whole concern about this approach would be there is no guarantee that user values are serializable...

As you are considering options, you may want to look into the upcoming `recoil-sync` library: https://github.com/facebookexperimental/Recoil/pull/1462/files

**Reply by nobleach** [reply]
ah, I think I missed the fact that using `getNodes_UNSTABLE()` would actually contain the values as well as the key names. I had console.logged out the result of that but never actually iterated it! Thank you for the heads up on recoil-sync. I'll have a look.

**Reply by nobleach** [reply]
It doesn't appear to contain values, just keys.
```
    const snapshot = useRecoilSnapshot();
    const nodeSnapshot = snapshot.getNodes_UNSTABLE();
    console.log('nodeSnapshot:', nodeSnapshot.next().value);
```
yields:
```
nodeSnapshot: RecoilState { key: 'productId' }
```
I'd expect there to be a value in there somewhere.... should there be?

**Reply by drarmstr** [reply]
Right -- it iterates the keys, then you can use `getLoadable()` or `getInfo_UNSTABLE()` to get the values.

**Reply by nobleach** [reply]
`getLoadable()` doesn't take the key name as a string though, it looks like it takes a pointer to the actual atom. (as in, I'll need to import the atom wherever I'm iterating).  Let me try `getInfo_UNSTABLE()`

**Reply by drarmstr** [reply]
`getLoadable()` takes an atom or selector as the parameter, not just the key.  But, this is what `getNodes_UNSTABLE()` provides.  So you can iterate that and pass the node directly to the accessor methods.

**Reply by nobleach** [reply]
I SEE! Ok, this is making far more sense to me. Thank you!
```
    const snapshot = useRecoilSnapshot();
    const nodeSnapshot = snapshot.getNodes_UNSTABLE();
    for (const a of nodeSnapshot) {
        const val = snapshot.getLoadable(a).contents;
        console.log('val:', val);
    }
```
is giving me exactly what I'd expect. Now I just need to call this AFTER all transactions are complete.

Thanks again!
